File: mobile/www/js/core/api/anilibria.js
// Бесплатный публичный API AniLibria: каталог, релизы, расписание. При сетевом сбое пробуем зеркало.
import { NetworkError } from "../errors.js";
import { request } from "./http.js";

const MIRRORS = ["https://anilibria.top", "https://api.anilibria.app"];
let mirror = 0;
export const site = () => MIRRORS[mirror];

/** Сроки свежести (секунды): справочники живут долго, ссылки на видео не кэшируются. */
export const TTL = { REFERENCES: 86400, FRANCHISE: 86400, CATALOG: 600, SEARCH: 600, SCHEDULE: 900, LATEST: 300,
  RELEASE: 120, PLAYER_SOURCE: 0 };

export async function al(path, params, ttl = 0, signal) {
  for (let i = 0; i < MIRRORS.length; i++) {
    try {
      return await request(`${site()}/api/v1${path}`, { params, ttl, signal });
    } catch (e) {
      if (!(e instanceof NetworkError) || i === MIRRORS.length - 1) throw e;
      mirror = (mirror + 1) % MIRRORS.length;
    }
  }
}

export function mediaUrl(path) {
  if (!path) return null;
  return path.startsWith("http") ? path : site() + path;
}

export function posterUrl(rel) {
  const p = rel?.poster || {};
  return mediaUrl(p.optimized?.src || p.src || p.preview);
}

export const catalog = (f = {}, signal) => al("/anime/catalog/releases", {
  page: f.page || 1, limit: f.limit || 30, "f[search]": f.search, "f[genres]": (f.genres || []).join(",") || null,
  "f[types]": f.types?.length ? f.types : null, "f[years][from_year]": f.yearFrom, "f[years][to_year]": f.yearTo,
  "f[sorting]": f.sorting, "f[publish_statuses]": f.ongoing == null ? null : (f.ongoing ? "IS_ONGOING" : "IS_NOT_ONGOING"),
  "f[seasons]": f.season,
}, TTL.CATALOG, signal);

export const latest = (signal) => al("/anime/releases/latest", { limit: 24 }, TTL.LATEST, signal);
// fresh — без кэша: ссылки на видео AniLibria привязаны к сети (VPN/страна) и после её смены не работают
export const release = (id, fresh = false, signal) => al(`/anime/releases/${id}`, null, fresh ? TTL.PLAYER_SOURCE : TTL.RELEASE, signal);
export const schedule = (signal) => al("/anime/schedule/week", null, TTL.SCHEDULE, signal);
export const genres = () => al("/anime/catalog/references/genres", null, TTL.REFERENCES);
export const years = () => al("/anime/catalog/references/years", null, TTL.REFERENCES);
export const franchiseAL = (id) => al(`/anime/franchises/release/${id}`, null, TTL.FRANCHISE);
export const searchAL = (q) => al("/app/search/releases", { query: q }, TTL.SEARCH);

/** Видео для замера скорости: самое высокое качество последней серии свежего релиза. */
export function sampleStream(releases) {
  for (const r of releases || []) {
    const ep = r.latest_episode || {};
    for (const q of ["1080", "720", "480"]) if (ep[`hls_${q}`]) return ep[`hls_${q}`];
  }
  return null;
}

File: anime_app/infrastructure/api/anilibria.py
"""Бесплатный публичный API AniLibria (AniLiberty): каталог, релизы, расписание. Ключи не нужны."""
from __future__ import annotations

from ...core.config import ANILIBRIA_MIRRORS, TTL
from ...core.errors import NetworkError
from ..http.client import HttpClient, RequestHandle, RequestScope

TYPES = [
    ("TV", "ТВ"), ("ONA", "ONA"), ("WEB", "WEB"), ("OVA", "OVA"), ("OAD", "OAD"),
    ("MOVIE", "Фильм"), ("DORAMA", "Дорама"), ("SPECIAL", "Спешл"),
]
SORTINGS = [
    ("FRESH_AT_DESC", "Недавно обновлённые"),
    ("RATING_DESC", "По рейтингу"),
    ("YEAR_DESC", "Сначала новые"),
    ("YEAR_ASC", "Сначала старые"),
    ("FRESH_AT_ASC", "Давно обновлённые"),
]
SEASONS = [("winter", "Зима"), ("spring", "Весна"), ("summer", "Лето"), ("autumn", "Осень")]


class AniLibriaApi:
    def __init__(self, http: HttpClient):
        self.http = http
        self._mirror = 0

    @property
    def site(self) -> str:
        return ANILIBRIA_MIRRORS[self._mirror]

    def media_url(self, path):
        """Постеры и превью приходят относительными путями."""
        if not path:
            return None
        if path.startswith("http"):
            return path
        return self.site + path

    def poster_url(self, release: dict | None, size: str = "src"):
        poster = (release or {}).get("poster") or {}
        opt = poster.get("optimized") or {}
        return self.media_url(opt.get(size) or poster.get(size) or poster.get("src"))

    # ------------------------------------------------------------------ core
    def get(self, path, params=None, on_ok=None, on_err=None, cache_ttl=0, scope: RequestScope | None = None,
            _attempt=0) -> RequestHandle:
        """Сетевой сбой (не ответ сервера) — пробуем зеркало."""
        def fail(err):
            if isinstance(err, NetworkError) and _attempt < len(ANILIBRIA_MIRRORS) - 1:
                self._mirror = (self._mirror + 1) % len(ANILIBRIA_MIRRORS)
                self.get(path, params, on_ok, on_err, cache_ttl, scope, _attempt + 1)
            elif on_err:
                on_err(err)
        return self.http.request(f"{self.site}/api/v1{path}", params, on_ok, fail, cache_ttl=cache_ttl, scope=scope)

    # ------------------------------------------------------------- endpoints
    def latest(self, on_ok, on_err=None, limit=24):
        return self.get("/anime/releases/latest", {"limit": limit}, on_ok, on_err, cache_ttl=TTL.LATEST)

    def release(self, id_or_alias, on_ok, on_err=None, fresh=False, scope=None):
        # fresh=True — без кэша: ссылки на видео привязаны к сети (VPN/страна), старые перестают работать
        return self.get(f"/anime/releases/{id_or_alias}", None, on_ok, on_err,
                        cache_ttl=TTL.PLAYER_SOURCE if fresh else TTL.RELEASE, scope=scope)

    def catalog(self, on_ok, on_err=None, page=1, limit=30, search=None, genre=None,
                types=None, year_from=None, year_to=None, sorting=None, genres=None,
                ongoing=None, season=None, scope=None):
        """Каталог AniLibria. genres — список id (нужны все выбранные), ongoing — True/False/None."""
        genre_ids = list(genres or []) + ([genre] if genre else [])
        params = {
            "page": page,
            "limit": limit,
            "f[search]": search,
            "f[genres]": ",".join(str(g) for g in genre_ids) or None,
            "f[types]": types,
            "f[years][from_year]": year_from,
            "f[years][to_year]": year_to,
            "f[sorting]": sorting,
            "f[publish_statuses]": None if ongoing is None else ("IS_ONGOING" if ongoing else "IS_NOT_ONGOING"),
            "f[seasons]": season,
        }
        return self.get("/anime/catalog/releases", params, on_ok, on_err, cache_ttl=TTL.CATALOG, scope=scope)

    def genres(self, on_ok, on_err=None):
        return self.get("/anime/catalog/references/genres", None, on_ok, on_err, cache_ttl=TTL.REFERENCES)

    def years(self, on_ok, on_err=None):
        return self.get("/anime/catalog/references/years", None, on_ok, on_err, cache_ttl=TTL.REFERENCES)

    def schedule(self, on_ok, on_err=None):
        return self.get("/anime/schedule/week", None, on_ok, on_err, cache_ttl=TTL.SCHEDULE)

    def franchise(self, release_id, on_ok, on_err=None):
        return self.get(f"/anime/franchises/release/{release_id}", None, on_ok, on_err, cache_ttl=TTL.FRANCHISE)

    def search(self, query, on_ok, on_err=None):
        return self.get("/app/search/releases", {"query": query}, on_ok, on_err)


def sample_stream(releases: list[dict]) -> str | None:
    """Видео для замера скорости: самое высокое качество последней серии свежего релиза."""
    for r in releases or []:
        ep = r.get("latest_episode") or {}
        for q in ("1080", "720", "480"):
            if ep.get(f"hls_{q}"):
                return ep[f"hls_{q}"]
    return None
